### PR TITLE
Confidential Ledger "Short LRO"

### LINKED ISSUE #21026: [FEATURE REQ] Add WaitForCompletion to Operation & Operation<T>
**Library or service name.**
Azure.Core
 
**Is your feature request related to a problem? Please describe.**
There is interest from the Synapse and Management teams for a built in and supported WaitForCompletion on Operation & Operation<T>

Today clients are writing their own, and with the non-generic Operation two copes are required. It's also pretty easy to introduce sync over async issues by mistake.

cc @pakrym

### PR DESCRIPTION
**Pull request by christothes**
The focus of these changes is to implement the "short LRO" pattern and plumb through support into `Azure.Core` for sync `WaitForCompletion` and `WaitForCompletionResponse` operations.

closes #21026

### CODE REVIEW

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/Azure.Security.ConfidentialLedger.csproj
Comment (pakrym) [inline sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/Azure.Security.ConfidentialLedger.csproj:18]: It just did.

File: sdk/core/Azure.Core/tests/samples/LoggingSamples.cs
Comment (pakrym) [inline sdk/core/Azure.Core/tests/samples/LoggingSamples.cs:91]: Thank you, I think we can remove these tests altogether.

File: sdk/core/Azure.Core/src/OperationOfT.cs
Comment (pakrym) [inline sdk/core/Azure.Core/src/OperationOfT.cs:60]: This is a good change but some people might be against it. Might be better to have a separate PR for it.

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/PostLedgerEntryOperation.cs
Comment (pakrym) [inline sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/PostLedgerEntryOperation.cs:47]: Use the OperationInternal for the implementation. (https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Shared/OperationInternal.cs)

It will do all the right things for you.

Comment (pakrym) [reply sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/PostLedgerEntryOperation.cs:47]: cc @kinelski

Comment (christothes) [reply sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/PostLedgerEntryOperation.cs:47]: cool - Does this make sense? https://github.com/Azure/azure-sdk-for-net/issues/22426

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/Azure.Security.ConfidentialLedger.csproj
Comment (christothes) [reply sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/Azure.Security.ConfidentialLedger.csproj:18]: Not with the changes in this PR :)

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/PostLedgerEntryOperation.cs
Comment (pakrym) [reply sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/PostLedgerEntryOperation.cs:47]: Good idea.

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/ConfidentialLedgerClient.cs
Comment (maririos) [inline sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/ConfidentialLedgerClient.cs:79]: should the `UpdateStatus` method be added here too and change the docstring to something like... "If false, use blah or blah to poll the results to ensure the operation completes"

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/tests/ConfidentialLedgerClientLiveTests.cs
Comment (maririos) [inline sdk/confidentialledger/Azure.Security.ConfidentialLedger/tests/ConfidentialLedgerClientLiveTests.cs:169]: just curious, if the default for `waitForCompletion` is always true, why do u pass it?

Comment (maririos) [reply sdk/confidentialledger/Azure.Security.ConfidentialLedger/tests/ConfidentialLedgerClientLiveTests.cs:169]: is it because of LLC stuff?

File: sdk/core/Azure.Core/src/OperationFailedException.cs
Comment (maririos) [inline sdk/core/Azure.Core/src/OperationFailedException.cs:34]:
```suggestion
     
        /// <param name="message">The error message that explains the reason for the exception.</param>
        /// <param name="operationId">The Id of the <see cref="Operation{T}"/> related to this exception.</param>
        /// <param name="innerException">The exception that is the cause of the current exception, or a null reference (Nothing in Visual Basic) if no inner exception is specified.</param>
```

File: sdk/core/Azure.Core/src/OperationOfT.cs
Comment (maririos) [reply sdk/core/Azure.Core/src/OperationOfT.cs:60]: is this something that could be added to Azure.Core.Experimental as a way to test and get feedback?

File: sdk/core/Azure.Core/src/OperationFailedException.cs
Comment (maririos) [inline sdk/core/Azure.Core/src/OperationFailedException.cs:20]: In which cases will the `OperationId` be null?

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/PostLedgerEntryOperation.cs
Comment (maririos) [inline sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/PostLedgerEntryOperation.cs:95]: any chance there could be a test added for this scenario?

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/tests/ConfidentialLedgerClientLiveTests.cs
Comment (christothes) [reply sdk/confidentialledger/Azure.Security.ConfidentialLedger/tests/ConfidentialLedgerClientLiveTests.cs:169]: Partly - because the tests can see the internal overload, the arg is required to disambiguate the two.

File: sdk/core/Azure.Core/src/OperationFailedException.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/OperationFailedException.cs:20]: There may not be a valid case.

File: sdk/core/Azure.Core/src/OperationOfT.cs
Comment (pakrym) [reply sdk/core/Azure.Core/src/OperationOfT.cs:60]: Interesting idea. We might be able to ship it as an extension method in experimental.

Comment (christothes) [reply sdk/core/Azure.Core/src/OperationOfT.cs:60]: Are there any specific concerns with shipping it in Core?

Comment (pakrym) [reply sdk/core/Azure.Core/src/OperationOfT.cs:60]: Not from me. Get an API approval early though.

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/api/Azure.Security.ConfidentialLedger.netstandard2.0.cs
Comment (KrzysztofCwalina) [inline sdk/confidentialledger/Azure.Security.ConfidentialLedger/api/Azure.Security.ConfidentialLedger.netstandard2.0.cs:36]: This is very long. I wonder if OperationPollingInterval would not be good enough name. Such value would be shared between all LROs on the client, but maybe it's OK?

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/api/Azure.Security.ConfidentialLedger.netstandard2.0.cs
Comment (KrzysztofCwalina) [inline sdk/confidentialledger/Azure.Security.ConfidentialLedger/api/Azure.Security.ConfidentialLedger.netstandard2.0.cs:30]: We need to decide if we keep using the Start prefix for SLROs.

Comment (christothes) [reply sdk/confidentialledger/Azure.Security.ConfidentialLedger/api/Azure.Security.ConfidentialLedger.netstandard2.0.cs:30]: In previous discussions, we were leaning towards not using the Start prefix. Are there some new thoughts around keeping it?

Comment (KrzysztofCwalina) [reply sdk/confidentialledger/Azure.Security.ConfidentialLedger/api/Azure.Security.ConfidentialLedger.netstandard2.0.cs:30]: I got confused by the usage samples. Given the default for waitForCompletion is true, I think we should not use the Start prefix. The usage samples pass true explicitly and so I thought the default is not waiting.

File: sdk/core/Azure.Core/src/OperationOfT.cs
Comment (KrzysztofCwalina) [reply sdk/core/Azure.Core/src/OperationOfT.cs:60]: We should review it at the arch meeting and once approved change our design guidelines.

File: sdk/core/Azure.Core/src/OperationFailedException.cs
Comment (kinelski) [inline sdk/core/Azure.Core/src/OperationFailedException.cs:30]: nit: doubled `summary` tags.

File: sdk/core/Azure.Core/src/Shared/OperationHelpers.cs
Comment (kinelski) [inline sdk/core/Azure.Core/src/Shared/OperationHelpers.cs:72]: Would it make more sense if we just called the sync `UpdateStatus` directly?

File: sdk/core/Azure.Core/src/Shared/OperationHelpers.cs
Comment (kinelski) [inline sdk/core/Azure.Core/src/Shared/OperationHelpers.cs:113]: Same here. Should we call the sync `UpdateStatus` instead?

File: sdk/core/Azure.Core/src/Shared/OperationHelpers.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/Shared/OperationHelpers.cs:72]: Good catch!

File: sdk/core/Azure.Core/src/Shared/OperationHelpers.cs
Comment (kinelski) [inline sdk/core/Azure.Core/src/Shared/OperationHelpers.cs:66]: The `WaitForCompletionAsync` implementation in `OperationInternal` uses the service's `retry-after` header (when available) to set the polling interval. Should we do the same in these `DefaultWaitForCompletion` implementations?

File: sdk/core/Azure.Core/src/Operation.cs
Comment (kinelski) [inline sdk/core/Azure.Core/src/Operation.cs:94]: I could be missing something, but why are the async `WaitForCompletion` methods abstract, while we provide default implementations to the sync ones?

File: sdk/core/Azure.Core/src/Shared/OperationHelpers.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/Shared/OperationHelpers.cs:66]: Looking at this again, none of the wait methods are ever referenced, so I think they aren't needed anymore.

File: sdk/core/Azure.Core/src/Operation.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/Operation.cs:94]: To avoid a breaking change.

File: sdk/core/Azure.Core/tests/OperationInternalTests.cs
Comment (christothes) [reply sdk/core/Azure.Core/tests/OperationInternalTests.cs]: I mean the CreateOperationOfT method must return `OperationInternal` - so even when `OperationInternalOfT` is returned, it uses the base implementation.

File: sdk/core/Azure.Core/src/Shared/OperationInternal.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/Shared/OperationInternal.cs:183]: But in the case that WaitForCompletion would call this internal method instead of the one that returns OperationState<T> and not set `Value`?

https://github.com/Azure/azure-sdk-for-net/blob/c5f840df810a9017cf597f6b21bfd84251e875c5/sdk/core/Azure.Core/src/Shared/OperationInternal.cs#L233

File: sdk/core/Azure.Core/src/Shared/OperationHelpers.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/Shared/OperationHelpers.cs]: I've refactored the wait logic to OperationHelpers now.

File: sdk/core/Azure.Core/src/Operation.cs
Comment (kinelski) [reply sdk/core/Azure.Core/src/Operation.cs:94]: I could be wrong, but I think that changing a method from abstract to virtual is not a breaking change. If I'm not missing any scenario, any implementation that overrides the abstract method would work if we changed the method to virtual. We're just allowing derived classes to omit the implementation from now on. In [Types of breaking changes](https://docs.microsoft.com/en-us/dotnet/core/compatibility/#members), it also mentions that:
> ✔️ ALLOWED: Changing a member from abstract to virtual

I don't really feel strongly about it, I just think that it could make future operations a bit shorter and with less repeated code. Maybe this is a discussion for another PR.

Comment (christothes) [reply sdk/core/Azure.Core/src/Operation.cs:94]: Ah good point - I was thinking about it the opposite way (adding an additional abstract method)

File: sdk/core/Azure.Core/src/Shared/OperationInternal.cs
Comment (pakrym) [reply sdk/core/Azure.Core/src/Shared/OperationInternal.cs:184]: Can we make it virtual an override in `OperationInternal<T>` to do the right thing?

File: sdk/core/Azure.Core/src/Operation.cs
Comment (pakrym) [inline sdk/core/Azure.Core/src/Operation.cs:68]: Any reason not to inline the DefaultWaitForCompletionResponseAsync implementation?

File: sdk/core/Azure.Core/src/Shared/OperationHelpers.cs
Comment (pakrym) [inline sdk/core/Azure.Core/src/Shared/OperationHelpers.cs:85]: Hm, it's not a new bug but we are ingoring retry-after header.

File: sdk/core/Azure.Core/api/Azure.Core.net461.cs
Comment (pakrym) [inline sdk/core/Azure.Core/api/Azure.Core.net461.cs:141]: Just a though. Can we make this one virtual as well?

Comment (pakrym) [reply sdk/core/Azure.Core/api/Azure.Core.net461.cs:141]: In the future.

**Review by pakrym — APPROVED** [review]

**Review by pakrym — APPROVED** [review]

**Pull request merged**
